### PR TITLE
Add iiif-crop to spotlight

### PR DESCRIPTION
**Pull request by jcoyne**
Things preventing this from being merged:
- [x] https://github.com/sul-dlss/exhibits/issues/311
- [x] https://github.com/projectblacklight/spotlight/issues/1570
- [x] https://github.com/sul-dlss/iiif-cropper/issues/40
- [x] https://github.com/projectblacklight/spotlight/issues/1569
- [x] https://github.com/projectblacklight/spotlight/issues/1545
- [x] https://github.com/projectblacklight/spotlight/issues/1648

Nice to have:

- [ ] https://github.com/sul-dlss/exhibits/issues/309
- [x] https://github.com/projectblacklight/spotlight/issues/1535

### CODE REVIEW

**Comment by cbeer** [comment]
Rebased.

File: app/assets/javascripts/spotlight/croppable.js [hunk 1]
```diff
@@ -4,115 +4,19 @@ Spotlight.onLoad(function() {
...
 /*
-  Croppable plugin
-  Implements http://deepliquid.com/content/Jcrop.html
-  Add jcrop data-attributes to file input (with data-croppable='true') to instantiate.
-  Adds initialSetSelect option to set a select box on the intial upload of an object.
+  IIIF image cropping plugin
+  Add iiif-crop data-attributes to file input (with data-croppable='true') to instantiate.
 */
 
+var xosd;  // TODO debug -- remove.
```
Comment (jkeck) [inline app/assets/javascripts/spotlight/croppable.js]: I'll kill this.

File: app/assets/javascripts/spotlight/search_typeahead.js [hunk 2]
```diff
@@ -59,7 +59,8 @@ function addAutocompletetoMastheadUpload(){
     }).on('typeahead:selected typeahead:autocompleted', function(e, data) {
       var remoteUrlField = $($(this).data('remoteUrlField'));
       var panel = $($(this).data('target-panel'));
-      swapInputForPanel($(this), panel, data);
+      // swapInputForPanel($(this), panel, data);
```
Comment (jkeck) [inline app/assets/javascripts/spotlight/search_typeahead.js]: I'll kill this cruft.

File: app/assets/stylesheets/spotlight/_croppable.scss [hunk 3]
```diff
@@ -1,16 +1,109 @@
...
+  left: -6px;
+  position: absolute;
+}
+
+.iiif-crop-bottom-right-drag-node {
+  bottom: -6px;
+  cursor: se-resize;
+  position: absolute;
+  right: -6px;
+}
```
Comment (jkeck) [inline app/assets/stylesheets/spotlight/_croppable.scss]: Just realized most of this CSS is useless now that we've swapped croppers.  I'll remove this.

File: app/controllers/spotlight/appearances_controller.rb [hunk 4]
```diff
@@ -22,20 +27,34 @@ def edit
 
     protected
 
+    def update_masthead
+      @exhibit.masthead ||= Spotlight::Masthead.new
+      @exhibit.masthead.update(params.require(:exhibit).require(:masthead_attributes).permit(featured_image_params))
+      @exhibit.save
```
Comment (jkeck) [inline app/controllers/spotlight/appearances_controller.rb]: This `masthead ||= Spotlight::Masthead.new` pattern shows up in a few places.  I feel like there is a better rails-y way to do this that is escaping me right now.

File: app/controllers/spotlight/contact_images_controller.rb [hunk 5]
```diff
@@ -0,0 +1,12 @@
+module Spotlight
+  # Handles requests to upload images for contact images
+  class ContactImagesController < FeaturedImagesController
```
Comment (jkeck) [inline app/controllers/spotlight/contact_images_controller.rb]: Can we accomplish this by routing the appropriate path to the `FeaturedImagesController` and making sure that `avatar_attributes` are handles properly there?

File: app/controllers/spotlight/mastheads_controller.rb [hunk 6]
```diff
@@ -0,0 +1,16 @@
+module Spotlight
+  # Handles requests to upload images for exhibit and site masthead images
+  class MastheadsController < FeaturedImagesController
```
Comment (jkeck) [inline app/controllers/spotlight/mastheads_controller.rb]: Same question here wrt the possibility of simply using the `FeaturedImagesController`.

File: app/models/spotlight/featured_image.rb [hunk 7]
```diff
@@ -4,16 +4,22 @@ module Spotlight
   class FeaturedImage < ActiveRecord::Base
     mount_uploader :image, Spotlight::FeaturedImageUploader
 
-    before_validation :set_image_from_uploaded_resource
+    # before_validation :set_image_from_uploaded_resource
```
Comment (jkeck) [inline app/models/spotlight/featured_image.rb]: I believe this was only necessary under carrierwave.  Things seem to work without it.  I'll remove this line as well as the method below.

File: app/views/spotlight/sir_trevor/blocks/_featured_pages_block.html.erb [hunk 8]
```diff
@@ -3,7 +3,7 @@
     <% featured_pages_block.pages.each_with_index do |page, index| %>
       <div class="box category-<%= (index + 1) %>">
         <%= link_to [page.exhibit, page] do %>
-          <div class="browse-category" style='background-image: linear-gradient(rgba(0, 0, 0, 0.0), rgba(0, 0, 0, 0.1), rgba(0, 0, 0, 0.5)) <%= ", url(\"#{page.thumbnail.image.thumb}\")" if page.thumbnail %>'>
+          <div class="browse-category" style='background-image: linear-gradient(rgba(0, 0, 0, 0.0), rgba(0, 0, 0, 0.1), rgba(0, 0, 0, 0.5)) <%= ", url(\"#{riiif.image_path(page.thumbnail, size: Spotlight::Engine.config.featured_image_thumb_size.join(','))}\")" if page.thumbnail %>'>
```
Comment (jkeck) [inline app/views/spotlight/sir_trevor/blocks/_featured_pages_block.html.erb]: This should probably be using the thumbnail's `iiif_url` instead of this (which won't account for crop coordinates)

File: app/controllers/spotlight/featured_images_controller.rb [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+    private
+
+    def tilesource
+      riiif.info_url(@featured_image.id)
+    end
+
+    # The create action can be called from a number of different forms, so
+    # we normalize all the parameters.
+    # rubocop:disable Metrics/CyclomaticComplexity, Metrics/MethodLength, Metrics/PerceivedComplexity
+    def create_params
```
Comment (cbeer) [inline app/controllers/spotlight/featured_images_controller.rb]: Is it possible to normalize these in the various forms?

Comment (jkeck) [inline app/controllers/spotlight/featured_images_controller.rb]: Yeah, I'll see if I can just set a static name on the file input.  We'll need to support at least two keys (since thumbnails and mastheads can exist in the same form, and it would be odd to have two file inputs w/ the same name in them.

**Comment by jkeck** [comment]
BTW, the reduction in coverage is due to the fact that we have builds for two versions of ruby with slightly different coverage metrics.

coverage


The -/+ 0.1% coverage comes from the discrepancy between those two builds and which one completes first.

File: app/controllers/spotlight/appearances_controller.rb (hunk 4 above)
Comment (cbeer) [inline app/controllers/spotlight/appearances_controller.rb]: `masthead = @exhibit.masthead || @exhibit.build_masthead` might be a little better

File: app/controllers/spotlight/appearances_controller.rb [hunk 10]
```diff
@@ -7,6 +7,11 @@ class AppearancesController < Spotlight::ApplicationController
 
     def update
       if @exhibit.update(exhibit_params)
+        # Update masthead attributes, only after we have saved the masthead_id
+        update_masthead
```
Comment (cbeer) [inline app/controllers/spotlight/appearances_controller.rb]: Is it possible to push this down into the model? Or, at least, leave a comment about why it isn't?

File: app/controllers/spotlight/featured_images_controller.rb [hunk 11]
```diff
@@ -0,0 +1,31 @@
+module Spotlight
+  # Handles requests to upload images for exhibit thumbnails
+  class FeaturedImagesController < Spotlight::ApplicationController
+    load_and_authorize_resource instance_name: :featured_image
+
+    def create
+      @featured_image.save!
```
Comment (cbeer) [inline app/controllers/spotlight/featured_images_controller.rb]: Are we relying on the exception for anything, or could this be written as something like: `if @featured_image.save && @featured_image.image.file`? (maybe with `@featured_image.image.file` behind a clearly named accessor?)

File: app/helpers/spotlight/exhibits_helper.rb [hunk 12]
```diff
@@ -0,0 +1,9 @@
+module Spotlight
+  # View helpers for the ExhibitsController
+  module ExhibitsHelper
+    # use the specified crop points, but a bigger image than the thumbnail
```
Comment (cbeer) [inline app/helpers/spotlight/exhibits_helper.rb]: Is this a TODO? Or old code?

File: app/models/concerns/spotlight/image_derivatives.rb [hunk 13]
```diff
@@ -1,57 +1,26 @@
...
-  #     version :my_version do
-  #       process :resize_to_fill => [30,30]
-  #     end
-  #   }
+  #   field: :my_field
   # }
-  #
-  # This will then create that new CarrierWave version in any class that extends this module
-  # and calls the apply_spotlight_image_derivative_versions class method described below.
   module ImageDerivatives
```
Comment (cbeer) [inline app/models/concerns/spotlight/image_derivatives.rb]: This module strikes me as somehow less useful now, yeah?

File: app/models/iiif_url.rb [hunk 14]
```diff
@@ -0,0 +1,31 @@
...
+      end
+
+      def #{name}=(val)
+        @parts['#{name}'] = val
+      end
+    EORUBY
+  end
+
+  def to_s
+    # TODO: prefix should be optional
```
Comment (cbeer) [inline app/models/iiif_url.rb]: Is this TODO ticketed?

File: app/models/concerns/spotlight/image_derivatives.rb (hunk 13 above)
Comment (jkeck) [inline app/models/concerns/spotlight/image_derivatives.rb]: There is definitely opportunity for refactoring, but I believe this is currently serving a purpose and can't quite be removed right now.  It provides the ability to configure the different solr fields that are used to populate the image references serialized into the sir-trevor block data.

I think when we take on #1673 (basically next) will be a good time to come up with a better way to configure which image versions are used in the application (I'll add a note in that ticket about doing so).

**Comment by jkeck** [comment]
Thanks for the helpful feedback @cbeer.  I believe I have addressed all but one of the suggestions, which I have commented about (`Spotlight::ImageDerivatives` refactoring).

Let me know if anything else is jumping out at you.

File: app/models/concerns/spotlight/image_derivatives.rb (hunk 13 above)
Comment (cbeer) [inline app/models/concerns/spotlight/image_derivatives.rb]: 👍

**Comment by tpendragon** [comment]
@jkeck @cbeer Chances this will get merged this week? And is there any sort of migration? Having a lot of call for mastheads lately, and this would be really nice to have.

**Comment by jkeck** [comment]
@tpendragon I hope so.  FWIW, there are a couple of `[WIP]` PRs that we will need to get reviewed/merged before this work is in a releasable state.  Mainly #1674 and #1675.  I also have progress on #1669 and will have a PR up today.

WRT a migration, we would also need to complete #1672.  There is some code in the `iiif_crop` branch, so #1672 is basically to ensure that it works given the refactoring we have done since it was originally written last summer.

**Comment by jkeck** [comment]
Also, FWIW, I think we can decouple the changes to the sir-trevor widgets (described in #1673) but I would be interested in opinions if we should ship a version of spotlight w/o changes to the sir-trevor widgets.  That may be a topic during the community call today.

File: app/views/spotlight/appearances/edit.html.erb [hunk 15]
```diff
@@ -31,15 +31,19 @@
     <div class="tab-content">
       <div role="tabpanel" class="tab-pane active" id="site-masthead">
         <p class="instructions"><%= t(:'.site_masthead.help') %></p>
+        <% image_form = iiif_cropper(f, :masthead, 1200, 120) %>
```
Comment (cbeer) [inline app/views/spotlight/appearances/edit.html.erb]: We should pull these values out into configuration.

File: app/controllers/spotlight/appearances_controller.rb [hunk 16]
```diff
@@ -23,19 +23,20 @@ def edit
     protected
 
     def exhibit_params
-      params.require(:exhibit).permit(main_navigations_attributes: [:id, :display, :label, :weight],
```
Comment (cbeer) [inline app/controllers/spotlight/appearances_controller.rb]: Can we move the `main_navigations_attributes` to the top of the list to avoid making this unnecessary change?

File: app/controllers/spotlight/contacts_controller.rb [hunk 17]
```diff
@@ -45,14 +47,15 @@ def attach_breadcrumbs
       add_breadcrumb t(:'spotlight.pages.index.about_pages.header'), exhibit_about_pages_path(@exhibit)
     end
 
+    def update_avatar
```
Comment (cbeer) [inline app/controllers/spotlight/contacts_controller.rb]: This method is now unused.

File: app/models/spotlight/exhibit.rb [hunk 18]
```diff
@@ -103,6 +107,18 @@ def reindex_progress
       @reindex_progress ||= ReindexProgress.new(resources) if resources
     end
 
+    def update_masthead(attributes = {})
+      to_be_updated = masthead || build_masthead
```
Comment (cbeer) [inline app/models/spotlight/exhibit.rb]: I think `update_masthead` and `update_thumbnail` are now unused?

File: app/models/spotlight/search.rb [hunk 19]
```diff
@@ -84,6 +86,18 @@ def merge_params_for_search(params, blacklight_config)
       base_query.params_for_search(user_query).merge(user_query.slice(:page))
     end
 
+    def update_masthead(attributes = {})
```
Comment (cbeer) [inline app/models/spotlight/search.rb]: I think `update_masthead` and `update_thumbnail` are now unused.

File: app/models/spotlight/search.rb [hunk 20]
```diff
@@ -71,6 +72,7 @@ def set_default_thumbnail
           )
         end
       end
+      save
```
Comment (cbeer) [inline app/models/spotlight/search.rb]: This `save` surprises me. Does it really belong here?

File: app/models/spotlight/site.rb [hunk 21]
```diff
@@ -13,5 +13,11 @@ class Site < ActiveRecord::Base
     def self.instance
       first || create
     end
+
```
Comment (cbeer) [inline app/models/spotlight/site.rb]: I think this method is unused.

File: app/services/spotlight/carrierwave_file_resolver.rb [hunk 22]
```diff
@@ -2,7 +2,10 @@ module Spotlight
   # Used by RIIIF to find files uploaded by carrierwave
   class CarrierwaveFileResolver < Riiif::AbstractFileSystemResolver
     def pattern(id)
-      Spotlight::FeaturedImage.find(id).image.file.file
+      # TODO: need to make Resource::Upload be a kind of FeaturedImage.
```
Comment (cbeer) [inline app/services/spotlight/carrierwave_file_resolver.rb]: Is this still a TODO? What's broken without it?

File: app/views/spotlight/pages/_form.html.erb [hunk 23]
```diff
@@ -61,9 +61,11 @@
 
       <% if @page.respond_to? :thumbnail %>
       <div role="tabpanel" class="tab-pane" id="page-thumbnail">
+        <% image_form = iiif_cropper(f, :thumbnail, *Spotlight::Engine.config.thumbnail_initial_crop_selection) %>
+        <%= image_form.hidden_field %>
```
Comment (cbeer) [inline app/views/spotlight/pages/_form.html.erb]: I think this field should be dropped `image_form.hidden_field`

File: app/views/spotlight/searches/_form.html.erb [hunk 24]
```diff
@@ -42,16 +42,20 @@
         <% end unless @search.query_params.blank? %>
       </div>
       <div role="tabpanel" class="tab-pane" id="search-masthead">
+        <% image_form = iiif_cropper(f, :masthead, *Spotlight::Engine.config.masthead_initial_crop_selection) %>
+        <%= image_form.hidden_field %>
```
Comment (cbeer) [inline app/views/spotlight/searches/_form.html.erb]: I think this field should be dropped `image_form.hidden_field`

File: app/views/spotlight/searches/_form.html.erb [hunk 25]
```diff
@@ -42,16 +42,20 @@
...
         <%= f.fields_for :masthead, (@search.masthead || @search.build_masthead) do |m| %>
           <p class="instructions"><%= t(:'.masthead.help') %></p>
           <p class="instructions"><%= t(:'.masthead.help_secondary') %></p>
-          <%= render '/spotlight/featured_images/form', f: m, jcrop_options: default_masthead_jcrop_options %>
+          <%= render '/spotlight/featured_images/form', f: m, image_form: image_form %>
         <% end %>
       </div>
       <div role="tabpanel" class="tab-pane" id="search-thumbnail">
+        <% image_form = iiif_cropper(f, :thumbnail, *Spotlight::Engine.config.featured_image_thumb_size) %>
+        <%= image_form.hidden_field %>
```
Comment (cbeer) [inline app/views/spotlight/searches/_form.html.erb]: I think this field should be dropped `image_form.hidden_field`

File: app/views/spotlight/sites/edit.html.erb [hunk 26]
```diff
@@ -20,8 +20,11 @@
 
         <div role="tabpanel" class="tab-pane" id="site-masthead">
           <p class="instructions"><%= t(:'.site_masthead.help') %></p>
+          <% image_form = iiif_cropper(f, :masthead, *Spotlight::Engine.config.masthead_initial_crop_selection) %>
+          <%= image_form.hidden_field %>
```
Comment (cbeer) [inline app/views/spotlight/sites/edit.html.erb]: I think this field should be dropped `image_form.hidden_field`

File: spec/factories/resources.rb [hunk 27]
```diff
@@ -6,6 +6,7 @@
   end
   factory :uploaded_resource, class: Spotlight::Resources::Upload, parent: :resource do
     type 'Spotlight::Resources::Upload'
-    url { Rack::Test::UploadedFile.new(File.expand_path(File.join('..', 'fixtures', 'avatar.png'), __dir__)) }
+    association :upload, factory: :featured_image
+    # url { Rack::Test::UploadedFile.new(File.expand_path(File.join('..', 'fixtures', 'avatar.png'), __dir__)) }
```
Comment (cbeer) [inline spec/factories/resources.rb]: Can we remove this commented-out code?

File: app/helpers/spotlight/crop_helper.rb [hunk 28]
```diff
@@ -0,0 +1,14 @@
+module Spotlight
+  ##
+  # iiif-crop options helpers
+  module CropHelper
+    def iiif_cropper(form, name, width, height)
```
Comment (cbeer) [inline app/helpers/spotlight/crop_helper.rb]: This is somewhat... dissatisfying.

File: app/controllers/spotlight/contacts_controller.rb [hunk 29]
```diff
@@ -10,6 +10,8 @@ class ContactsController < Spotlight::ApplicationController
 
     def new
       add_breadcrumb t(:'helpers.action.spotlight/contact.create'), new_exhibit_contact_path(@exhibit)
+      # Enable the nested form to be drawn.
```
Comment (cbeer) [inline app/controllers/spotlight/contacts_controller.rb]: I wonder if this is still needed?

**Comment by cbeer** [comment]
Should we update to leaflet 1.0.3?

File: app/controllers/spotlight/contacts_controller.rb (hunk 29 above)
Comment (jkeck) [inline app/controllers/spotlight/contacts_controller.rb]: Yeah, I think you're right.  Running tests w/ this removed to verify.
